src/libostree/ostree-repo.c

```diff
@@ -4379,6 +4379,261 @@ ostree_repo_read_commit (OstreeRepo   *self,
+
+  switch (type)
+    {
+    case G_FILE_TYPE_DIRECTORY:
+      if (!(flags & OSTREE_REPO_OPEN_FILE_EXPECT_TREE))
+        {
+          g_autofree char * basename = g_file_get_basename ((GFile*) ret_out);
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_REGULAR_FILE,
+                       "Bad filespec \"%s\": \"%s\" is a directory, not a file",
+                       ref, basename);
+          return FALSE;
+        }
+      break;
+    case G_FILE_TYPE_REGULAR:
+    case G_FILE_TYPE_SYMBOLIC_LINK:
+      if (!(flags & OSTREE_REPO_OPEN_FILE_EXPECT_FILE))
+        {
+          g_autofree char * basename = g_file_get_basename ((GFile*) ret_out);
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_DIRECTORY,
+                       "Bad treeish \"%s\": \"%s\" is a file, not a directory",
+                       ref, basename);
+          return FALSE;
+        }
+      break;
+    case G_FILE_TYPE_UNKNOWN:
+      {
+        g_autofree char * basename = g_file_get_basename ((GFile*) ret_out);
+        g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                     "Bad ref \"%s\": \"%s\": No such file or directory",
+                     ref, basename);
+        return FALSE;
+      }
+      break;
+    default:
+      g_assert_not_reached ();
+    }
+
+  ot_transfer_out_value (out, &ret_out);
+  ot_transfer_out_value (out_commit_sha, &commit_sha);
+  return TRUE;
+}
+
+static gboolean
+resolve_sha_to_tree (OstreeRepo       *self,
+                     const char*       sha,
+                     OstreeRepoFile  **out,
+                     char            **out_commit_sha,
+                     GCancellable     *cancellable,
+                     GError           **error)
+{
+  gboolean has_object;
+  if (!ostree_repo_has_object (self, OSTREE_OBJECT_TYPE_COMMIT, sha,
+                               &has_object, cancellable, error))
+    return FALSE;
+
+  if (has_object)
+    {
+      g_autoptr(GFile) root_gfile = NULL;
+      if (!ostree_repo_read_commit (self, sha, &root_gfile, NULL, cancellable,
+                                    error))
+        return FALSE;
+      g_autoptr(OstreeRepoFile) root = OSTREE_REPO_FILE (root_gfile);
+      g_return_val_if_fail (root, FALSE);
+      root_gfile = NULL;
+      if (out_commit_sha)
+        *out_commit_sha = g_strdup (sha);
+      ot_transfer_out_value (out, &root);
+      return TRUE;
+    }
+
+  if (!ostree_repo_has_object (self, OSTREE_OBJECT_TYPE_DIR_TREE, sha,
+                               &has_object, cancellable, error))
+    return FALSE;
+  if (has_object)
+    {
+      /* We're a bit cheeky here - we need to specify a metadata checksum but we
+       * don't have one.  We're getting a subpath here anyway we're not going to
+       * return this invalid file. */
+      g_autoptr(OstreeRepoFile) root =  _ostree_repo_file_new_root (
+          self, sha, "00000000000000000000000000000000");
```

Comment:
This is where I set the "00000000000000000000000000000000" SHA.  I guess the reference above deserves a comment.